Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean

```diff
@@ -101,6 +129,36 @@ alias ⟨BlockTriangular.of_charmatrix, BlockTriangular.charmatrix⟩ := charmat
 def charpoly (M : Matrix n n R) : R[X] :=
   (charmatrix M).det
 
+@[simp]
+theorem charpoly_zero : charpoly (0 : Matrix n n R) = X ^ Fintype.card n := by
+  simp [charpoly]
+
+theorem charpoly_diagonal (d : n → R) : charpoly (diagonal d) = ∏ i, (X - C (d i)) := by
+  simp [charpoly]
+
+
+theorem charpoly_one : charpoly (1 : Matrix n n R) = (X - 1) ^ Fintype.card n := by
+  simp [charpoly]
+
+theorem charpoly_natCast (k : ℕ) :
+    charpoly (k : Matrix n n R) = (X - (k : R[X])) ^ Fintype.card n := by
+  simp [charpoly]
+
+theorem charpoly_ofNat (k : ℕ) [k.AtLeastTwo] :
+    charpoly (ofNat(k) : Matrix n n R) = (X - ofNat(k)) ^ Fintype.card n:=
+  charpoly_natCast _
+
+theorem eval_charpoly (A : Matrix m m R) (r : R) :
```

Comment:
nit: I think this file is easier to follow if this lemma is placed just before cayley-hamilton, since all the others are about the polynomial (or polynomial matrix) itself whereas this is about evaluation of the polynomial, like cayley-hamilton is

Reply:
I'd argue the opposite; this belongs immediately after the definition, to support the docstring

> The characteristic polynomial of a matrix `M` is given by $\det (t I - M)$.

Reply:
(I guess I should rename my variables to match)

Reply:
I'm happy with it immediately after the definition too, my real confusion was that it was in the middle of an otherwise reasonable chain

Reply:
I've moved it and added the `LinearMap` version too.